src/commands/publish/upload_form/project_assets.rs

```diff
@@ -0,0 +1,43 @@
+use super::binding::Binding;
+use super::file::File;
+use super::kv_namespace::KvNamespace;
+use super::wasm_module::WasmModule;
+
+#[derive(Debug)]
+pub struct ProjectAssets {
+    pub script_path: String,
+    pub wasm_modules: Vec<WasmModule>,
+    pub kv_namespaces: Vec<KvNamespace>,
+}
+
+impl ProjectAssets {
+    pub fn files(&self) -> Vec<File> {
+        let mut files = Vec::new();
+        let script = File {
+            name: "script".to_string(),
+            path: self.script_path.clone(),
+        };
+        files.push(script);
+
+        for wm in &self.wasm_modules {
+            let file = wm.file();
+            files.push(file);
+        }
+
+        files
+    }
+
+    pub fn bindings(&self) -> Vec<Binding> {
+        let mut bindings = Vec::new();
+        for wm in &self.wasm_modules {
+            let binding = wm.binding();
+            bindings.push(binding);
+        }
+        for kv in &self.kv_namespaces {
+            let binding = kv.binding();
+            bindings.push(binding);
+        }
+
+        bindings
+    }
```

Comment:
I really want to write this as

```rust
pub fn bindings(&self) -> Vec<Binding> {
    self.wasm_modules
        .iter()
        .chain(self.kv_namespaces.iter())
        .map(|e| e.binding())
        .collect()    
}
```

but, that doesn't work since they're different types.

You *could* write

```rust
pub fn bindings(&self) -> Vec<Binding> {
    self.wasm_modules
        .iter()
        .map(|m| m.binding())
        .chain(self.kv_namespaces.iter().map(|n| n.binding())
        .collect()    
}
```

but I think this is stretching the limits of Too Many Iterators :)

Another option is

```rust
    let mut bindings: Vec<_> = wasm_modules.iter().map(|m| m.binding()).collect();
    bindings.extend(kv_namespaces.iter().map(|n| n.binding()));
    bindings
```

... but after considering all of this, i think the original code is probably the most clear in this case. I guess.

Reply:
😂